### PR TITLE
Fix support for Ctrl-C on the macosx backend.

### LINKED ISSUE #3991: SIGINT is ignored by MacOSX backend
Here is an example:

```
import matplotlib
#matplotlib.use('Qt4Agg')
from matplotlib import pyplot as plt
plt.plot(1, 1)
plt.show()
```

Run this script from terminal, wait for the window to appear and press Ctrl-C (in terminal). The script will not die.
Then uncomment the `matplotlib.use` line an Ctrl-C will work as intended.

UPD:
If I press Ctrl-C and then move mouse over the plot window the following exception appears in  the log:

```
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.3/lib/python3.3/site-packages/matplotlib/backend_bases.py", line 1850, in enter_notify_event
    def enter_notify_event(self, guiEvent=None, xy=None):
KeyboardInterrupt
```

The same is for callback set using `signal.signal(signal.SIGTERM, cb)`, callback is not called until window receives any event (focus change, mouse move, etc).

### LINKED ISSUE #10002: can't stop macosx mainloop
### Bug report

**Bug summary**

I can't figure out how to stop the cocoa mainloop once it's going with the macosx backend. This could be a missing API, or just one that I don't know about.

I suspect an API that exposes sending `[NSApp stop]` is all that's needed, but I'm not sure.

**Code for reproduction**

```python
import matplotlib
matplotlib.use('macosx')
import matplotlib.pyplot as plt
from matplotlib.backends.backend_macosx import TimerMac, show

def stop_mainloop():
    print("I don't know how to quit you")
    t.stop()

t = TimerMac(2000)
t.add_callback(stop_mainloop)
t.start()

show.mainloop()
```

**Actual outcome**

Loop runs forever because I don't know how to stop it.

**Matplotlib version**

  * Operating system: macOS 10.13.2
  * Matplotlib version: 2.0.2 (conda-forge)
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.6.1

### PR DESCRIPTION
**Pull request by anntzer**
Support is largely copy-pasted from, and tests are shared with, the qt implementation (qt_compat._maybe_allow_interrupt, #13306), the main difference being that what we need from QSocketNotifier, as well as the equivalent for QApplication.quit(), are reimplemented in ObjC.

qt_compat._maybe_allow_interrupt is also slightly cleaned up by moving out the "do-nothing" case (`old_sigint_handler in (None, SIG_IGN, SIG_DFL)`) and dedenting the rest, instead of keeping track of whether signals were actually manipulated via a `skip` variable.

Factoring out the common parts of _maybe_allow_interrupt is left as a follow-up.

(Test e.g. with
`MPLBACKEND=macosx python -c "from pylab import *; plot(); show()"` followed by Ctrl-C.)

Closes #3991 (which was closed as cantfix in https://github.com/matplotlib/matplotlib/issues/3991#issuecomment-271174177; likely much of #4006 can now be reverted).

Also closes #10002 by providing _macosx.stop as a private function to call [NSApp stop]; whether we want to make that public is another question.

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by greglucas** [comment]
Nice, I think this could simplify a lot of the objc code, and I think this PR should probably clean up while you are working on this. See here for a quick hammer to your linked issue that seems to still work as expected for me. https://github.com/matplotlib/matplotlib/commit/5a30956189808138027daddc67ffcfba16a72b3b
Feel free to take any or all of that yourself.

I also get the test failure timeout locally. If I click on the app icon to bring the window back into focus then the test passes. So, my guess is there is something with the event loop going to sleep and not being woken back up again?

Also just as a note since I was looking at this portion of the code here. I am pretty sure we've got some leaks in the window objects on the objc side (loop over `plot(); show();` in memory_profiler and watch it build up) , and now that I'm seeing the retains in the `WindowServerConnectionManager` I'm wondering if some of that shared connection handling is keeping things alive and that would be worth looking at.

File: lib/matplotlib/tests/test_backend_qt.py [hunk 1]
```diff
@@ -53,201 +53,6 @@ def test_fig_close():
...
-        proc.wait_for('SIGUSR1')
-        os.kill(proc.pid, signal.SIGINT)
-        stdout, _ = proc.communicate(timeout=_test_timeout)
-    except Exception:
-        proc.kill()
-        stdout, _ = proc.communicate()
-        raise
-    print(stdout)
-    assert 'SUCCESS' in stdout
-    plt.figure()
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_qt.py:191]: @QuLogic Do you remember why you added this plt.figure() line here (in #20907?)  AFAICT it is not needed for the test and also locks in the running interactive framework for the main process, so I'm going to remove it(?)

**Comment by anntzer** [comment]
Thanks for the quick review.  I just force-pushed a tiny cleanup for now as well as the change to the tests mentioned just above, I'll integrate your large cleanup patch later once we decide what to do with the problem below.

I can repro the failure on test_other_signal_before_sigint, with the same observation as you.  Doing a bit of printf-debugging (e.g. adding a printf at the beginning of receivedData, and also a print in the _WaitForStringPopen.wait_for loop) suggests to me that there may(?) be a race condition between set_wakeup_fd and PyErr_CheckSignals: it seems like receivedData sometimes gets called before PyErr_CheckSignals will actually call a signal handler.  Somewhat curiously, adding the printf also makes that test pass *sometimes*, though not always, which supports the interpretation that there's a bit of delay needed before we can usefully call PyErr_CheckSignals (and the tiny delay from the printf helps making that call successful).
If we can't figure that out quickly, I'd suggest just xfailing that test for now (perhaps on a shorter timeout...), as the behavior improvement on the plain test_sigint seems worth it by itself.

File: lib/matplotlib/tests/test_backend_qt.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/tests/test_backend_qt.py:191]: I'm not seeing any reason for it to be there either right now, and don't recall anything specific.

**Comment by greglucas** [comment]
Interestingly, I have had a similar experience debugging some previous issues with the reference counting where I added in some `NSLog()` statements, and that miraculously fixed the segfaults I was trying to locate. I would agree though, if it isn't quick to find something reliable then xfailing seems reasonable, or just removing the macosx from this list of backends because we aren't testing anything other than qt currently anyways. Another thing to try would be to call `flush_events()` to try and spin the event loop on the objc side.

**Comment by anntzer** [comment]
Calling flush_events (or rather inlining its implementation into receivedData) doesn't seem to help, so I've just removed the test for now.

**Review by greglucas — APPROVED** [review]
This handles the KeyboardInterrupt as I'd expect now when testing manually on macosx and qt.

It would be nice to figure out what is going on with the multiple signals test case on macosx, but I don't think that should hold things up right now.

Some more comments for future selves:
https://docs.python.org/3/c-api/exceptions.html#c.PyErr_CheckSignals
> If the function is called from a non-main thread, or under a non-main Python interpreter, it does nothing and returns 0.

Are we possibly calling this from a different background NSApp thread? since you aren't specifying which queue to put the notification on?

When I run `pytest -n 4 lib/matplotlib/tests/test_backends_interactive.py::test_other_signal_before_sigint` and don't do anything. I get both tests to fail, `[pause-kwargs1-macosx]` and `[show-kwargs0-macosx]`. If I move my mouse over the figure (must activate NSTrackingRect?) during the test, the `[show-kwargs0-macosx]` test succeeds, but the `[pause-kwargs1-macosx]` fails, so there must be a difference between show and pause in the interaction here too.

Do we need to drain the filehandle/read what was there already and again call `waitForDataInBackgroundAndNotify`?

**Comment by anntzer** [comment]
Using mainQueue doesn't seem to help (but it may well be that it's indeed needed but not enough).
Perhaps we indeed need to repeat the wait but then I'm not sure why the printfs would sometimes help, and also I'd rather just defer implementing that to later (unless you see an easy way to do it).

**Comment by greglucas** [comment]
No I don't see anything immediately obvious, my approval still stands, I was just throwing out ideas. It's just fine to leave that for later IMO.

**Comment by jklymak** [comment]
Is the test failure unrelated?  I haven't dug into this, but would be happy to merge on Greg's recommendation.  Except the MacOS test is failing...

**Comment by ksunden** [comment]
Closing and reopening to rerun tests with current main.

I'm pretty sure the test failure is the tk problems on azure that we were having for a while (and actually only skips still as we await azure fixing things on their end). Though the run is old enough to not be accessible anymore, so can't be fully sure.

**Comment by greglucas** [comment]
Should be good to go from my perspective and CI is now passing, but note that there is some minor QT refactoring going on here as well.

File: lib/matplotlib/tests/test_backends_interactive.py [hunk 2]
```diff
@@ -613,3 +614,118 @@ def test_figure_leak_20490(env, time_mem):
...
+@pytest.mark.parametrize("backend", ["qtagg", "macosx"])
+@pytest.mark.parametrize("target, kwargs", [
+    ('show', {'block': True}),
+    ('pause', {'interval': 10})
+])
+def test_sigint(backend, target, kwargs):
+    try:
+        plt.switch_backend(backend)
+    except ImportError as exc:
+        pytest.skip(f"Failed to switch to backend {backend} ({exc}).")
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/tests/test_backends_interactive.py]: sure, fixed.

File: lib/matplotlib/tests/test_backends_interactive.py [hunk 3]
```diff
@@ -631,3 +632,116 @@ def test_figure_leak_20490(env, time_mem):
...
+@pytest.mark.parametrize("backend", ["qtagg"])
+@pytest.mark.parametrize("target, kwargs", [
+    ('show', {'block': True}),
+    ('pause', {'interval': 10})
+])
+def test_other_signal_before_sigint(backend, target, kwargs):
+    try:
+        plt.switch_backend(backend)
+    except ImportError as exc:
+        pytest.skip(f"Failed to switch to backend {backend} ({exc}).")
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backends_interactive.py:730]: There's the same backend switch here.

Comment (anntzer) [reply lib/matplotlib/tests/test_backends_interactive.py]: oops, sorry, fixed again.

**Comment by greglucas** [comment]
@anntzer, this needs a rebase now.

**Comment by anntzer** [comment]
Thanks for the bump, rebased.

**Comment by ksunden** [comment]
Failing test is one of the newly added tests

**Comment by greglucas** [comment]
> Failing test is one of the newly added tests

I can confirm this test fails locally for me as well. It is only the `show` + `block=True` combination that fails though which is odd. I think this combo can get skipped/xfailed for now with a comment indicating it is a known issue and if someone has time to investigate it then they can. But, this PR overall is a net gain and that can be saved for later IMO. My approval still stands.

**Comment by greglucas** [comment]
I dug into this a bit more and it seems like the `show()` version isn't starting the event loop right away, but instead sitting idly by and waiting for something else to wake it up. If you run the test and don't touch the mouse it fails, but if you move the mouse over the canvas that is enough to wake it up and trigger the events and make the test pass. So, I think we need to somehow explicitly start an event loop when we go through `show()`. `pause()` explicitly calls `canvas.start_event_loop(interval)`, which is why this test passes for `pause()`.

**Comment by anntzer** [comment]
I'll have limited access to a dev machine for a few days, so feel free to push an xfail for now or a fix.

**Comment by anntzer** [comment]
Added the xfail for the show() + other signal case.  At least from a quick look show(block=True) should call start_main_loop (via pyplot_show) so it's unclear why the event loop would not be started?

**Pull request merged**